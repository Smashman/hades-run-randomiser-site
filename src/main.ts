import { Run, RunOptions } from './run';

const runOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        heatRange: {
            min: 4,
            max: 12,
        },
    }
};

const run = Run.generateRandomRun(runOptions);

console.log('run', run);
console.log('run.weapon', run.weapon);
console.log('run.weaponAspect', run.weaponAspect);
console.log('run.keepsake', run.keepsake);
console.log('run.companion', run.companion);
console.log('run.mirror', run.mirror);
console.log('run.pact', run.pact);