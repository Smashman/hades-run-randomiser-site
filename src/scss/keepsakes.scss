@use 'variables' as v;

$black: black;
$grey: #0a0b0d;
$blue: #677a88;
$gold: #8b6439;

table.display-case {
    $left-right-border: 9px solid $gold;
    $top-bottom-border: 7px solid $gold;

    background: $grey;
    border-collapse: collapse;
    empty-cells: show;
    border-top: $top-bottom-border;
    border-right: $left-right-border;
    border-bottom: $top-bottom-border;
    border-left: $left-right-border;
    tr {
        $border: 4px solid $gold;
        border: $border;
    }
    td {
        $width: 120px;
        $height: 158px;

        width: $width;
        height: $height;
        min-width: $width;
        min-height: $height;
        max-width: $width;
        max-height: $height;
        background-image: linear-gradient(to top, $black 1px, $blue 1px);
        padding: 0;
    }
}

.keepsake-cubby {
    $left-blue-width: 2px;
    $gradient-stop: 80px;
    $background-gradient: linear-gradient(to right, $blue $left-blue-width, transparent $left-blue-width), linear-gradient(to bottom, $grey $gradient-stop, $blue $gradient-stop);
    $border-radius: 10px;

    background-image: $background-gradient;
    width: 100%;
    height: 100%;
    border-radius: 0 0 $border-radius $border-radius;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-direction: column;
    padding: 8px 2px;
    border: 1px solid $black;

    $icon-glow: drop-shadow(0 0 0.75em $gold);

    &:hover {
        &.hidden > div {
            opacity: 0.2;
        }
    }

    &.locked {
        .keepsake-backing:hover {
            .keepsake-icon {
                filter: opacity(0.5) $icon-glow;
            }
        }
    }
    
    .keepsake-name {
        cursor: default;
        @include v.alegreya-sans-sc-bold;
        color: white;
        font-size: 1.1em;
        line-height: 21px;
        height: 42px;
        text-align: center;
        display: flex;
        align-items: center;
        overflow: hidden;
    }

    .keepsake-backing {
        $size: 110px;

        width: $size;
        height: $size;
        background-image: url(v.$img-head-path + 'keepsake/backing.png');
        background-size: $size;
        background-repeat: no-repeat;

        &:hover {
            .keepsake-icon {
                filter: $icon-glow;
            }
        }

        .keepsake-icon {
            width: $size;
            height: $size;
        }
    }

    &.locked {
        .keepsake-icon {
            filter: opacity(0.5);
        }
    }

    &.hidden > div {
        opacity: 0;
    }
}

