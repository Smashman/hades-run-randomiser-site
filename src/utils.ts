export const getRandomItemFromArray = <T>(array: T[]) => array[Math.floor(Math.random() * array.length)];
export const fillArray = <T>(length: number, fillValue: T) => Array.from<T>({length: length}).fill(fillValue);

export interface Options {
    enabled?: boolean;
}

export abstract class Item {
    isUnlocked: boolean = false;

    unlock() {
        this.isUnlocked = true;
    }

    lock() {
        this.isUnlocked = false;
    }
}

export abstract class ItemList<T extends Item> {
    items: T[] = [];

    addItem(item: T) {
        this.items = [...this.items, item];
        return this;
    }

    addItems(items: T[]) {
        items.forEach(item => this.addItem(item));
        return this;
    }

    get unlocked() {
        return this.items.filter(item => item.isUnlocked);
    }

    unlockAll() {
        this.items.forEach(item => item.unlock());
    }

    getRandom(): T {
        return getRandomItemFromArray(this.unlocked) ?? null;
    }
}

export class Level {
    constructor(protected _value: number, readonly min: number, readonly max: number) {}

    set value(setTo: number) {
        if (setTo >= this.max) {
            this._value = this.max;
        }
        else if (setTo <= this.min) {
            this._value = this.min;
        }
        else {
            this._value = setTo;
        }
    }

    get value() {
        return this._value;
    }

    isMinLevel(): boolean {
        return this.value <= this.min;
    }

    isMaxLevel(): boolean {
        return this.value >= this.max;
    }

    setMaxLevel(): void {
        this.value = this.max;
    }

    addValue(change: number): number {
        this.value = this.value + change;
        return this.value;
    }

    increment(): number {
        return this.addValue(1);
    }

    decrement(): number {
        return this.addValue(-1);
    }
}