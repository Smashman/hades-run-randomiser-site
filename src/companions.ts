import { getRandomItemFromArray, Item, Options } from './utils';
import { ItemList } from './utils';

export class Companion extends Item {
    level: number = 1;
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level === Companion.maxLevel;
    }
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level = Companion.maxLevel);
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}