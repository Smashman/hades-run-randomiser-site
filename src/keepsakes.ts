import { getRandomItemFromArray, Options, ItemList, Item, Level } from './utils';

export interface KeepsakeOptions extends Options {
    ignoreMaxLevel?: boolean;
    changeBetweenBiomes?: boolean;
}

export class Keepsakes extends ItemList<Keepsake> {
    getRandomKeepsakes(keepsakeOptions?: KeepsakeOptions): KeepsakeConfiguration {
        const keepsakes = (keepsakeOptions?.ignoreMaxLevel && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(keepsake => !keepsake.isMaxLevel)] : [...this.unlocked];

        if (keepsakes.length === 0) {
            return null;
        }

        let keepsakesChosen: Keepsake[] = [getRandomItemFromArray(keepsakes)];

        if (keepsakeOptions?.changeBetweenBiomes) {
            for (let i = 1; i <= 3; i++) {
                keepsakesChosen = [...keepsakesChosen, getRandomItemFromArray(keepsakes.filter(keepsake => !keepsakesChosen.includes(keepsake))) ?? null];
            }
        }

        return keepsakesChosen as KeepsakeConfiguration;
    }

    maxLevelAll() {
        this.items.forEach(keepsake => keepsake.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(keepsake => keepsake.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(keepsake => keepsake.isMaxLevel);
    }
}

export class Keepsake extends Item {
    level: Level = new Level(1, Keepsake.minLevel, Keepsake.maxLevel);
    constructor(public name: string, public giver: string, public icon: string, public isHidden: boolean = false) {
        super();
    }

    static minLevel = 1;
    static maxLevel = 3;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }
}

export type KeepsakeConfiguration = [Keepsake] | [Keepsake, Keepsake, Keepsake, Keepsake] | null;