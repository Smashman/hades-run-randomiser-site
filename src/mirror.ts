import { getRandomItemFromArray, Item, Options, StorableItemData } from './utils';

export interface MirrorOptions extends Options {
};

export const defaultMirrorOptions: MirrorOptions = {
    enabled: true,
};

export class Mirror {
    groups: MirrorGroup[] = [];

    addGroup(sets: [[MirrorTalent, MirrorTalent], [MirrorTalent, MirrorTalent]], isUnlocked: boolean = false) {
        const mirrorTalentSets = <[MirrorTalentSet, MirrorTalentSet]>sets.map(set => new MirrorTalentSet(set));
        this.groups = [...this.groups, new MirrorGroup(mirrorTalentSets, isUnlocked)];
        return this;
    }

    get unlockedGroups() {
        return this.groups.filter(group => group.isUnlocked);
    }

    unlockAll() {
        this.groups.forEach(group => group.unlock());
    }

    getRandom(): MirrorConfiguration {
        return new MirrorConfiguration(this.unlockedGroups.map(group => group.getRandom()).reduce((acc, val) => [...acc, ...val], <MirrorTalent[]>[]));
    }

    toStorableData() {
        return {
            groups: this.groups.map(group => group.toStorableData())
        }
    }
}

export class MirrorGroup extends Item{
    constructor(public sets: [MirrorTalentSet, MirrorTalentSet], public isUnlocked: boolean = false) {
        super();
    }

    getRandom() {
        return <[MirrorTalent, MirrorTalent]>this.sets.map(set => set.getRandom());
    }
}

export class MirrorTalentSet {
    constructor(public talents: [MirrorTalent, MirrorTalent]) {
        talents[0].colour = "red";
        talents[1].colour = "green";
    }

    getRandom(): MirrorTalent {
        return getRandomItemFromArray(this.talents);
    }
}

export class MirrorTalent {
    rankCount: number;
    colour: "red" | "green";
    totalCost: number;

    constructor(public name: string, public rankCosts: number[]) {
        this.rankCount = rankCosts.length;
        this.totalCost = rankCosts.reduce((totalCost, cost) => totalCost + cost);
    }
}

export class MirrorConfiguration {
    totalDarkness: number;
    constructor(public talents: MirrorTalent[]){
        this.totalDarkness = talents.reduce((totalCost, talent) => totalCost + talent.totalCost, 0);
    };
}