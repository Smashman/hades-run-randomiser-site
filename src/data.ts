import { Weapon, WeaponAspect, Weapons } from './weapons';
import { Keepsake, Keepsakes } from './keepsakes';
import { Companion, Companions } from './companions';
import { Mirror, MirrorTalent } from './mirror';
import { Pact, PactCondition } from './pact';
import { fillArray } from './utils';
import * as React from 'react';
import * as keepsakeIcons from './img/keepsake';
import * as weaponIcons from './img/weapon';

export const weapons = new Weapons().addItems([
    new Weapon('Stygius', 'Blade of the Underworld', 'sword', true)
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.sword.zagreus, false, true),
        new WeaponAspect('Nemesis', weaponIcons.sword.nemesis),
        new WeaponAspect('Poseidon', weaponIcons.sword.poseidon),
        new WeaponAspect('Arthur', weaponIcons.sword.arthur, true),
    ]),
    new Weapon('Varatha', 'Eternal Spear', 'spear')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.spear.zagreus, false, true),
        new WeaponAspect('Achilles', weaponIcons.spear.achilles),
        new WeaponAspect('Hades', weaponIcons.spear.hades),
        new WeaponAspect('Guan Yu', weaponIcons.spear.guanYu, true),
    ]),
    new Weapon('Aegis', 'Shield of Chaos', 'shield')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.shield.zagreus, false, true),
        new WeaponAspect('Chaos', weaponIcons.shield.chaos),
        new WeaponAspect('Zeus', weaponIcons.shield.zeus),
        new WeaponAspect('Beowulf', weaponIcons.shield.beowulf, true),
    ]),
    new Weapon('Coronacht', 'Heart-Seeking Bow', 'bow')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.bow.zagreus, false, true),
        new WeaponAspect('Chiron', weaponIcons.bow.chiron),
        new WeaponAspect('Hera', weaponIcons.bow.hera),
        new WeaponAspect('Rama', weaponIcons.bow.rama, true),
    ]),
    new Weapon('Malphon', 'Twin Fists', 'fists')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.fists.zagreus, false, true),
        new WeaponAspect('Talos', weaponIcons.fists.talos),
        new WeaponAspect('Demeter', weaponIcons.fists.demeter),
        new WeaponAspect('Gilgamesh', weaponIcons.fists.gilgamesh, true),
    ]),
    new Weapon('Exagryph', 'Adamant Rail', 'gun')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.gun.zagreus, false, true),
        new WeaponAspect('Eris', weaponIcons.gun.eris),
        new WeaponAspect('Hestia', weaponIcons.gun.hestia),
        new WeaponAspect('Lucifer', weaponIcons.gun.lucifer, true),
    ]),
]);

export const keepsakes = new Keepsakes().addItems([
    new Keepsake('Old Spiked Collar', 'Cerberus', keepsakeIcons.oldSpikedCollar),
    new Keepsake('Myrmidon Bracer', 'Achilles', keepsakeIcons.myrmidonBracer),
    new Keepsake('Black Shawl', 'Nyx', keepsakeIcons.blackShawl),
    new Keepsake('Pierced Butterfly', 'Thanatos', keepsakeIcons.piercedButterfly),
    new Keepsake('Bone Hourglass', 'Charon', keepsakeIcons.boneHourglass),
    new Keepsake('Chthonic Coin Purse', 'Hypnos', keepsakeIcons.chthonicCoinPurse),
    new Keepsake('Skull Earring', 'Megaera', keepsakeIcons.skullEarring),
    new Keepsake('Distant Memory', 'Orpheus', keepsakeIcons.distantMemory),
    new Keepsake('Harpy Feather Duster', 'Dusa', keepsakeIcons.harpyFeatherDuster),
    new Keepsake('Lucky Tooth', 'Skelly', keepsakeIcons.luckyTooth),
    new Keepsake('Thunder Signet', 'Zeus', keepsakeIcons.thunderSignet),
    new Keepsake('Conch Shell', 'Poseidon', keepsakeIcons.conchShell),
    new Keepsake('Owl Pendant', 'Athena', keepsakeIcons.owlPendant),
    new Keepsake('Eternal Rose', 'Aphrodite', keepsakeIcons.eternalRose),
    new Keepsake('Blood-Filled Vial', 'Ares', keepsakeIcons.bloodFilledVial),
    new Keepsake('Adamant Arrowhead', 'Artemis', keepsakeIcons.adamantArrowhead),
    new Keepsake('Overflowing Cup', 'Dionysus', keepsakeIcons.overflowingCup),
    new Keepsake('Lambent Plume', 'Hermes', keepsakeIcons.lambentPlume),
    new Keepsake('Frostbitten Horn', 'Demeter', keepsakeIcons.frostbittenHorn),
    new Keepsake('Cosmic Egg', 'Chaos', keepsakeIcons.cosmicEgg),
    new Keepsake('Shattered Shackle', 'Sisyphus', keepsakeIcons.shatteredShackle),
    new Keepsake('Evergreen Acorn', 'Eurydice', keepsakeIcons.evergreenAcorn),
    new Keepsake('Broken Spearpoint', 'Patroclus', keepsakeIcons.brokenSpearpoint),
    new Keepsake('Pom Blossom', 'Persephone', keepsakeIcons.pomBlossom),
    new Keepsake('Sigil of the Dead', 'Hades', keepsakeIcons.sigilOfTheDead, true),
]);

export const companions = new Companions().addItems([
    new Companion('Battie', 'Megaera'),
    new Companion('Mort', 'Thanatos'),
    new Companion('Rib', 'Skelly'),
    new Companion('Shady', 'Sisyphus'),
    new Companion('Fidi', 'Dusa'),
    new Companion('Antos', 'Achilles'),
]);

export const mirror = new Mirror()
    .addGroup([
        [
            new MirrorTalent('Shadow Presence', [10, 15, 20, 25, 30]),
            new MirrorTalent('Fiery Presence', [10, 15, 20, 25, 30]),
        ], [
            new MirrorTalent('Chthonic Vitality', [10, 20, 40]),
            new MirrorTalent('Dark Regeneration', [30, 60]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Death Defiance', [30, 500, 1000]),
            new MirrorTalent('Stubborn Defiance', [600]),
        ],
        [
            new MirrorTalent('Greater Reflex', [50]),
            new MirrorTalent('Ruthless Reflex', [75]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Boiling Blood', [10, 30, 50, 70, 90]),
            new MirrorTalent('Abyssal Blood', [20, 40, 60, 80, 100]),
        ],
        [
            new MirrorTalent('Infernal Soul', [20, 80]),
            new MirrorTalent('Stygian Soul', [0, 60, 180]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Deep Pockets', [30, 35, 40, 45, 50, 55, 60, 65, 70, 75]),
            new MirrorTalent('Golden Touch', [70, 90, 110]),
        ],
        [
            new MirrorTalent('Thick Skin', [40, 45, 50, 55, 60, 65, 70, 75, 80, 85]),
            new MirrorTalent('High Confidence', [50, 100, 150, 200, 250]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Privileged Status', [50, 500]),
            new MirrorTalent('Family Favorite', [50, 150]),
        ],
        [
            new MirrorTalent('Olympian Favor', fillArray(40, 50)),
            new MirrorTalent('Dark Foresight', fillArray(10, 150)),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Gods\' Pride', fillArray(20, 100)),
            new MirrorTalent('Gods\' Legacy', fillArray(10, 250)),
        ],
        [
            new MirrorTalent('Fated Authority', [500, 750, 1000, 1250, 1500, 1750, 2000, 2250]),
            new MirrorTalent('Fated Persuasion', [1000, 2000, 3000, 4000]),
        ]
    ]);

export const pact = new Pact()
    .addCondition(new PactCondition('Hard Labor', [1, 1, 1, 1, 1]))
    .addCondition(new PactCondition('Lasting Consequences', [1, 1, 1, 1]))
    .addCondition(new PactCondition('Convenience Fee', [1, 1]))
    .addCondition(new PactCondition('Jury Summons', [1, 1, 1]))
    .addCondition(new PactCondition('Extreme Measures', [1, 2, 3, 4], 3))
    .addCondition(new PactCondition('Calisthenics Program', [1, 1]))
    .addCondition(new PactCondition('Benefits Package', [2, 3]))
    .addCondition(new PactCondition('Middle Management', [2]))
    .addCondition(new PactCondition('Underworld Customs', [2]))
    .addCondition(new PactCondition('Forced Overtime', [3, 3]))
    .addCondition(new PactCondition('Heightened Security', [1]))
    .addCondition(new PactCondition('Routine Inspection', [2, 2, 2, 2]))
    .addCondition(new PactCondition('Damage Control', [1, 1]))
    .addCondition(new PactCondition('Approval Process', [2, 3]))
    .addCondition(new PactCondition('Tight Deadline', [1, 2, 3]));

// weapons.unlockAll();
// keepsakes.unlockAll();
// companions.unlockAll();
// mirror.unlockAll();
// pact.unlockAll();

// weapons.maxLevelAll();
// keepsakes.maxLevelAll();
// companions.maxLevelAll();
// companions.maxCodexAll();

interface Data {
    weapons: Weapons;
    keepsakes: Keepsakes;
    companions: Companions;
    mirror: Mirror;
    pact: Pact;
}

export const defaultData: Data = {
    weapons,
    keepsakes,
    companions,
    mirror,
    pact,
};

export const DataContext = React.createContext<[Data, React.Dispatch<React.SetStateAction<Data>>]>([defaultData, {} as React.Dispatch<React.SetStateAction<Data>>]);

//@ts-ignore
// window.data = defaultData;