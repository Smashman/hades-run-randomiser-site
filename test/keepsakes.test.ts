import { Keepsake, Keepsakes } from '../src/keepsakes';
import type { KeepsakeConfiguration } from '../src/keepsakes';
import * as utils from '../src/utils';
import { getRandomItemFromArrayMock, fakeIcon } from './helpers';

describe('Keepsake class', () => {
    it('should create expected Keepsake class instance', () => {
        const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);

        expect(keepsake).toMatchSnapshot();
    });

    it('should return false for isMaxLevel when level is less than 3', () => {
        const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);
        keepsake.level.value = 2;

        expect(keepsake.isMaxLevel).toBe(false);
    });

    it('should return true for isMaxLevel when level is 3', () => {
        const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);
        keepsake.level.value = 3;

        expect(keepsake.isMaxLevel).toBe(true);
    });

    it('should set isHidden to true when passed into the constructor', () => {
        const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon, true);

        expect(keepsake.isHidden).toBe(true);
    });

    describe('toStorableData', () => {
        it('should return expected storable data', () => {
            const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);
    
            expect(keepsake.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with level', () => {
            const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);
            keepsake.level.value = 3;
    
            expect(keepsake.toStorableData()).toMatchSnapshot();
        });
    });

    describe('fromStoredData', () => {
        it('should retun expected Keepsake instance after loading from data', () => {
            const keepsake = new Keepsake('Shoe', 'Jeff', fakeIcon);
            keepsake.fromStoredData({
                isUnlocked: true,
                level: 2,
            });

            expect(keepsake).toMatchSnapshot();
        });
    });
});

describe('Keepsakes class', () => {
    it('should return false for isAllMaxLevel if not every keepsake is max level', () => {
        const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
        const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
        const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);

        keepsake1.level.value = 3;

        expect(keepsakes.isAllMaxLevel).toBe(false);
    });

    it('should return true for isAllMaxLevel if every keepsake is max level', () => {
        const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
        const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
        const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);

        keepsake1.level.value = 3;
        keepsake2.level.value = 3;

        expect(keepsakes.isAllMaxLevel).toBe(true);
    });

    it('should max level all keepsakes when maxLevelAll is called', () => {
        const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
        const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
        const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);

        keepsakes.maxLevelAll();

        expect(keepsake1.level.value).toBe(3);
        expect(keepsake2.level.value).toBe(3);

        expect(keepsakes.isAllMaxLevel).toBe(true);
    });

    describe('getRandomKeepsakes', () => {
        let getRandomItemFromArraySpy: jest.SpyInstance;

        let keepsakes: Keepsakes;

        let keepsake1: Keepsake;
        let keepsake2: Keepsake;
        let keepsake3: Keepsake;
        let keepsake4: Keepsake;

        beforeEach(() => {
            keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
            keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
            keepsake3 = new Keepsake('Skirt', 'Bob', fakeIcon);
            keepsake4 = new Keepsake('Scarf', 'Trevor', fakeIcon);
            keepsakes = new Keepsakes().addItems([keepsake1, keepsake2, keepsake3, keepsake4]);

            getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);
        });

        afterEach(() => {
            jest.spyOn(utils, 'getRandomItemFromArray').mockRestore();
        });

        it('should return null if no keepsakes are unlocked', () => {
            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes();

            expect(randomKeepsakeConfiguration).toBe(null);
            expect(getRandomItemFromArraySpy).not.toBeCalled();
        });

        it('should return KeepsakeConfiguration with one value when called with no options when all keepsakes are unlocked', () => {
            keepsakes.unlockAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes();

            const expectedChoices: Keepsake[] = [keepsake1, keepsake2, keepsake3, keepsake4];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake3];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should return KeepsakeConfiguration with one value when called with no options when a subset of keepsakes are unlocked', () => {
            keepsake2.unlock();
            keepsake4.unlock();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes();

            const expectedChoices: Keepsake[] = [keepsake2, keepsake4];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake4];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should return KeepsakeConfiguration with one value when called with changeBetweenBiomes option set to false when all keepsakes are unlocked', () => {
            keepsakes.unlockAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({changeBetweenBiomes: false});

            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake3];

            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should return KeepsakeConfiguration with four values when called with changeBetweenBiomes option set to true when all keepsakes are unlocked', () => {
            keepsakes.unlockAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({changeBetweenBiomes: true});

            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake3, keepsake2, keepsake4, keepsake1];

            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should return a keepsake configuration with 3 null values when called with changeBetweenBiomes option set to true if there\'s only one keepsake to choose from', () => {
            keepsakes = new Keepsakes().addItem(keepsake1);
            keepsakes.unlockAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({changeBetweenBiomes: true});

            const expectedChoices: Keepsake[] = [keepsake1];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake1, null, null, null];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should not filter out any keepsakes when called with ignoreMaxLevel true, but no keepsakes are at max level', () => {
            keepsakes.unlockAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({ignoreMaxLevel: true});

            const expectedChoices: Keepsake[] = [keepsake1, keepsake2, keepsake3, keepsake4];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake3];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should filter out keepsakes when called with ignoreMaxLevel true and keepsakes are at max level', () => {
            keepsakes.unlockAll();

            keepsake1.level.value = Keepsake.maxLevel;
            keepsake3.level.value = Keepsake.maxLevel;

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({ignoreMaxLevel: true});

            const expectedChoices: Keepsake[] = [keepsake2, keepsake4];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake4];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });

        it('should get random item from full array when called with ignoreMaxLevel true, but all keepsakes are at max level', () => {
            keepsakes.unlockAll();
            keepsakes.maxLevelAll();

            const randomKeepsakeConfiguration = keepsakes.getRandomKeepsakes({ignoreMaxLevel: true});

            const expectedChoices: Keepsake[] = [keepsake1, keepsake2, keepsake3, keepsake4];
            const expectedKeepsakeConfiguration: KeepsakeConfiguration = [keepsake3];

            expect(getRandomItemFromArraySpy).toBeCalledWith(expectedChoices);
            expect(randomKeepsakeConfiguration).toStrictEqual(expectedKeepsakeConfiguration);
        });
    });

    describe('toStorableData', () => {
        it('should return expected storable data', () => {
            const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
            const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
            const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);
    
            expect(keepsakes.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with level', () => {
            const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
            const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
            const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);

            keepsake1.level.value = 3;
    
            expect(keepsakes.toStorableData()).toMatchSnapshot();
        });
    });

    describe('fromStoredData', () => {
        it('should retun expected Keepsake instance after loading from data', () => {
            const keepsake1 = new Keepsake('Shoe', 'Jeff', fakeIcon);
            const keepsake2 = new Keepsake('Coat', 'Delilah', fakeIcon);
            const keepsakes = new Keepsakes().addItems([keepsake1, keepsake2]);
            keepsakes.fromStoredData({
                items: [{
                    isUnlocked: true,
                    level: 2,
                },
                {
                    isUnlocked: false,
                    level: 3,
                }]
            });

            expect(keepsakes).toMatchSnapshot();
        });
    });
});