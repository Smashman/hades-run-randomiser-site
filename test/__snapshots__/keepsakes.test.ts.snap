// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keepsake class fromStoredData should retun expected Keepsake instance after loading from data 1`] = `
Keepsake {
  "giver": "Jeff",
  "icon": "base64:icon!",
  "isHidden": false,
  "isUnlocked": true,
  "level": Level {
    "_value": 2,
    "max": 3,
    "min": 1,
    "onChange": undefined,
  },
  "name": "Shoe",
}
`;

exports[`Keepsake class should create expected Keepsake class instance 1`] = `
Keepsake {
  "giver": "Jeff",
  "icon": "base64:icon!",
  "isHidden": false,
  "isUnlocked": false,
  "level": Level {
    "_value": 1,
    "max": 3,
    "min": 1,
    "onChange": undefined,
  },
  "name": "Shoe",
}
`;

exports[`Keepsake class toStorableData should return expected storable data 1`] = `
Object {
  "isUnlocked": false,
  "level": 1,
}
`;

exports[`Keepsake class toStorableData should return expected storable data with level 1`] = `
Object {
  "isUnlocked": false,
  "level": 3,
}
`;

exports[`Keepsakes class toStorableData should return expected storable data 1`] = `
Object {
  "items": Array [
    Object {
      "isUnlocked": false,
      "level": 1,
    },
    Object {
      "isUnlocked": false,
      "level": 1,
    },
  ],
}
`;

exports[`Keepsakes class toStorableData should return expected storable data with level 1`] = `
Object {
  "items": Array [
    Object {
      "isUnlocked": false,
      "level": 3,
    },
    Object {
      "isUnlocked": false,
      "level": 1,
    },
  ],
}
`;
